internal/testing/require/kubernetes.go

```diff
@@ -0,0 +1,173 @@
+/*
+ Copyright 2021 - 2024 Crunchy Data Solutions, Inc.
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package require
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+	goruntime "runtime"
+	"strings"
+	"sync"
+	"testing"
+
+	"gotest.tools/v3/assert"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+
+	"github.com/crunchydata/postgres-operator/internal/controller/runtime"
+)
+
+// https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest#pkg-constants
+var envtestVarsSet = os.Getenv("KUBEBUILDER_ASSETS") != "" ||
+	strings.EqualFold(os.Getenv("USE_EXISTING_CLUSTER"), "true")
+
+// EnvTest returns an unstarted Environment with crds. It calls t.Skip when
+// the "KUBEBUILDER_ASSETS" and "USE_EXISTING_CLUSTER" environment variables
+// are unset.
+func EnvTest(t testing.TB, crds envtest.CRDInstallOptions) *envtest.Environment {
+	t.Helper()
+
+	if !envtestVarsSet {
+		t.SkipNow()
+	}
+
+	return &envtest.Environment{
+		CRDInstallOptions: crds,
+		Scheme:            crds.Scheme,
+	}
+}
+
+var kubernetes struct {
+	sync.Mutex
+
+	// Count references to the started Environment.
+	count int
+	env   *envtest.Environment
+}
+
+// Kubernetes starts or connects to a Kubernetes API and returns a client that uses it.
+// When starting a local API, the client is a member of the "system:masters" group.
+//
+// It calls t.Fatal when something fails. It stops the local API using t.Cleanup.
+// It calls t.Skip when the "KUBEBUILDER_ASSETS" and "USE_EXISTING_CLUSTER" environment
+// variables are unset.
+//
+// Tests that call t.Parallel might share the same local API. Call t.Parallel after this
+// function to ensure they share.
+func Kubernetes(t testing.TB) client.Client {
+	t.Helper()
+	_, cc := kubernetes3(t)
+	return cc
+}
+
+// Kubernetes2 is the same as [Kubernetes] but also returns a copy of the client
+// configuration.
+func Kubernetes2(t testing.TB) (*rest.Config, client.Client) {
+	t.Helper()
+	env, cc := kubernetes3(t)
+	return rest.CopyConfig(env.Config), cc
+}
+
+func kubernetes3(t testing.TB) (*envtest.Environment, client.Client) {
```

Comment:
Same here... :shrug:

Reply:
I started with this exported because a couple `setupKubernetes` had this shape. After seeing everything, it wasn't much to change those places to `Kubernetes2` and unexport this.

Reply:
I was glad to unexport it because I have no idea how to safely share an `envtest.Environment` value or what concurrency is okay on it. `Kubernetes2` returns a copy, which seems safe so far.